library/std/src/sys/pal/unix/fs.rs

```diff
@@ -739,44 +739,37 @@ impl Iterator for ReadDir {
                 // contents were "simply" partially initialized data.
                 //
                 // Like for uninitialized contents, converting entry_ptr to `&dirent64`
-                // would not be legal. However, unique to dirent64 is that we don't even
-                // get to use `&raw const (*entry_ptr).d_name` because that operation
-                // requires the full extent of *entry_ptr to be in bounds of the same
-                // allocation, which is not necessarily the case here.
-                //
-                // Instead we must access fields individually through their offsets.
-                macro_rules! offset_ptr {
-                    ($entry_ptr:expr, $field:ident) => {{
-                        const OFFSET: isize = mem::offset_of!(dirent64, $field) as isize;
-                        if true {
-                            // Cast to the same type determined by the else branch.
-                            $entry_ptr.byte_offset(OFFSET).cast::<_>()
-                        } else {
-                            #[allow(deref_nullptr)]
-                            {
-                                &raw const (*ptr::null::<dirent64>()).$field
-                            }
-                        }
+                // would not be legal. However, we can use `&raw const (*entry_ptr).d_name`
+                // to refer the fields individually, because that operation is equivalent
+                // to `byte_offset` and thus does not require the full extent of `*entry_ptr`
+                // to be in bounds of the same allocation, only the offset of the field
+                // being referenced.
+                macro_rules! entry_field_ptr {
+                    ($field:ident) => {{
+                        // To make sure the field actually exists and is visible,
+                        // and we aren't silently doing any Deref coercion.
+                        const _: usize = mem::offset_of!(dirent64, $field);
+                        &raw const (*entry_ptr).$field
```

Comment:
The deref concern makes sense to me (and is scary in general), but what happens if the first clause happens i.e. the field doesn't exist or isn't visible? Afaik, `&raw` doesn't allow bypassing field visibility.

Reply:
I think that is true, yes, any error about a field being missing or too private should cause an error on both the `offset_of!` and the `&raw` equally. I could remove the `the field actually exists and is visible, and` part of the comment if that would be clearer.

Reply:
I'm still not sure what the potential error is here - is the concern that `dirent64` may add a `Deref` implementation to something that also has a field named `d_ino`/`d_type`/ `d_name`? That seems unlikely enough that I don't see the need for checking `offset_of`. At least, I don't think this is any more of a concern for `dirent64` than for any other use of `&raw`.

@RalfJung any thoughts about this kind of pattern?

Reply:
Interesting, I hadn't considered the semver implications on this... moving a field behind a `Deref` impl is indeed a breaking change for some unsafe code users.

But I also never saw this pattern explicitly defending against that, not sure if it's worth it.